File: src/render.c
#include "render.h"
#include "common.h"
#include "player.h"
#include "wfc.h"
#include <asm-generic/ioctls.h>
#include <sys/ioctl.h>
#include <unistd.h>

void render_frame(void) {
  struct winsize ws;
  ioctl(STDOUT_FILENO, TIOCGWINSZ, &ws);

  int oy = (ws.ws_row - MAP_HEIGHT) / 2;
  int ox = (ws.ws_col - MAP_WIDTH) / 2;

  printf("\033[2J");

  for (int i = 0; i < MAP_HEIGHT * MAP_WIDTH; i++) {
    int y = i / MAP_WIDTH;
    int x = i % MAP_WIDTH;

    if (x == 0)
      printf("\033[%d;%dH", oy + y + 1, ox + 1);

    if (player.y == y && player.x == x) {
      printf("\033[38;5;226m@\033[0m");
    } else {
      TileType tile = map[y][x].final_type;
      const char *symbol;
      int color;

      if (tile == TILE_VOID) {
        symbol = " ";
        color = 232;
      } else if (tile == TILE_FLOOR) {
        symbol = "·";
        color = 82;
      } else if (tile == TILE_WALL) {
        symbol = "█";
        color = 196;
      } else if (tile == TILE_ALTAR) {
        symbol = "♦";
        color = 135;
      } else {
        symbol = "?";
        color = 255;
      }

      printf("\033[38;5;%dm%s\033[0m", color, symbol);
    }

    if (x == MAP_WIDTH - 1)
      printf("\n");
  }

  fflush(stdout);
}

File: src/render.rs
use crate::common::{MAP_HEIGHT, MAP_WIDTH};
use crate::player::Player;
use crate::wfc::{World, TILE_ALTAR, TILE_FLOOR, TILE_VOID, TILE_WALL};
use crossterm::terminal;
use std::io::{self, Write};

pub fn render_frame(world: &World, player: &Player) -> io::Result<()> {
    let (cols, rows) = terminal::size().unwrap_or((80, 24));
    let oy = (rows as i32 - MAP_HEIGHT as i32) / 2;
    let ox = (cols as i32 - (MAP_WIDTH * 2) as i32) / 2;

    let mut out = io::stdout().lock();
    write!(out, "\x1b[2J")?;

    for y in 0..MAP_HEIGHT {
        write!(out, "\x1b[{};{}H", oy + y as i32 + 1, ox + 1)?;
        for x in 0..MAP_WIDTH {
            if player.y == y && player.x == x {
                write!(out, "\x1b[38;5;226m@ \x1b[0m")?;
            } else {
                let tile = world.map[y][x].final_type;
                let (symbol, color) = match tile {
                    t if t == TILE_VOID => ("  ", 232),
                    t if t == TILE_FLOOR => ("· ", 82),
                    t if t == TILE_WALL => ("██", 196),
                    t if t == TILE_ALTAR => ("♦ ", 135),
                    _ => ("?", 255),
                };
                write!(out, "\x1b[38;5;{}m{}\x1b[0m", color, symbol)?;
            }
        }
    }

    out.flush()
}
